src/structs/parser.rs

```diff
@@ -0,0 +1,264 @@
+use crate::errors::*;
+use anyhow::{anyhow, Context, Result};
+use std::iter::Peekable;
+use Token::*;
+
+macro_rules! node {
+    ($token:expr) => {
+        node!($token, Box::new(None), Box::new(None))
+    };
+    ($token:expr, $left:expr) => {
+        node!($token, $left, Box::new(None))
+    };
+    ($token:expr, $left:expr, $right:expr) => {
+        Box::new(Some(Node::new($token, $left, $right)))
+    };
+}
+
+pub type Branch = Box<Option<Node>>;
+
+#[derive(Debug, Copy, Clone)]
+pub enum Direction {
+    UniDirectional,
+    BiDirectional,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum Token {
+    Implicator(Direction),
+    Operator(char),
+    Parenthesis(char),
+    Attribute(char),
+}
+
+#[derive(Debug)]
+pub struct Node {
+    _token: Token,
+    _left: Branch,
+    _right: Branch,
+}
+
+impl Node {
+    pub fn new(_token: Token, _left: Branch, _right: Branch) -> Node {
+        Node {
+            _token,
+            _left,
+            _right,
+        }
+    }
+}
+
+pub struct Parser;
+
+impl<'a> Parser {
```

Comment:
Fairly generic name that doesn't tell us much of what this does, same goes for the file name